EZP-29613: As a developer I want access ContentType on Content to avoid re-loading it

andrerom opened the pull request:
| Question      | Answer
| ------------- | ---
| Tickets       | [EZP-29613](https://jira.ez.no/browse/EZP-29613)
| New feature?  | yes
| BC breaks?    | no
| Tests pass?   | yes
| Doc needed?   | no
| License       | [GPL-2.0](https://github.com/ezsystems/ezplatform-admin-ui/blob/master/LICENSE)

Uses https://github.com/ezsystems/ezpublish-kernel/pull/2444, and `$location->getContent()`
(v2.2) in order avoid several repository loads not needed anymore.

Besides avoiding executing unneeded PHP code, here are some numbers for cache lookups on clean install, expect difference to be bigger when you have more data _(demo or project)_:

dashboard:
- 418 / 1240 (33.71%) _(master)_
- 468 / 1315 (35.59%) _(with kernel PR, only)_*
- 364 / 1043 (34.9%) _(with this PR + kernel PR)_

content structure:
- 478 / 1401 (34.12%)  _(master)_
- 450 / 1359 (33.11%) _(with kernel PR, only)_
- 436 / 1327 (32.86%) _(with this PR + kernel PR)_

_**EDIT: These numbers are now outdated, newer versions of admin-ui adds new load calls for among others users**_

_* So without this patch there is a slight regression with more cache loads with the kernel PR on dashboard. Probably due to kernel now always loading content types from API instead of falling back to using SPI, and via api language loading and mapping is needed. With this PR we get rid of the duplicated loads, and for other apps the extra language load and much much more will go away once we have inMemory cache again (v2.4)._

#### Checklist:
- [x] Coding standards (`$ composer fix-cs`)
- [x] Tests  **NOTE: There is a TMP commit here to get travis to test against kernel PR for it to pass**
- [x] Ready for Code Review

andrerom commented on src/lib/UI/Value/ValueFactory.php:
```diff
@@ -248,10 +249,8 @@ public function createBookmark(Location $location): UIValue\Location\Bookmark
-                    $this->locationService->loadLocation($location->contentInfo->mainLocationId)
-                ),
+                'contentType' => $location->getContent()->getContentType(),
+                'pathLocations' => $this->pathService->loadPathLocations($location),
```
NOTE: I assumed we don't need to load main location here but rather want path for the location bookmark is for.

andrerom commented on src/lib/UniversalDiscovery/Event/Subscriber/SectionAssign.php:
```diff
@@ -97,20 +96,15 @@ private function getRestrictedContentTypes(array $hasAccess): array
+            return [];
         }
 
-        $restrictedContentTypesIds = array_unique(array_merge(...$restrictedContentTypesIds));
```
Review Note: I didn't see why array_merge() was used here so removed it in the new code,

andrerom commented:
✅ Passing + kernel sent for QA. So this is ready for review now @ViniTou / @webhdx

lserwatka commented:
Note for QA: we only need sanities here. ping @micszo

alongosz commented on src/lib/EventListener/RequestAttributesListener.php:
```diff
@@ -63,7 +63,7 @@ public function addRequestAttributes(FilterViewBuilderParametersEvent $event)
 
         if ($parameterBag->has('locationId') && $request->get('_route') === '_ezpublishLocation') {
-            $location = $this->loadLocation($parameterBag->get('locationId'));
+            $location = $this->loadLocation($parameterBag->get('locationId'), $parameterBag->get('languageCode'));
```
@andrerom Are you sure `languageCode` is always present in the ParameterBag? There is `hasContentLanguage` method call right after that block of code, so I'm not certain.

andrerom replied:
Not 100% sure, just saw the code above and it will fallback to null if not set which fits.

andrerom replied:
Checked now, yeah this is intentional. 
In loadLocation  I pass it along as `$languageCode ? [$languageCode] : null`

andrerom commented:
@micszo That is due [to usage of non API (internal) properties](https://github.com/ezsystems/ezplatform-form-builder/blob/master/src/lib/Tab/LocationView/SubmissionsTab.php#L171-L173) on content object:
```
        $submissionsPagerfanta = new Pagerfanta(
            new SubmissionsAdapter($this->formSubmissionService, $content->contentInfo, $content->prioritizedFieldLanguageCode)
        );
```

`prioritizedFieldLanguageCode` is an internal property, and should not be used. As this PR instead injects a ContentProxy that property does not exists. Ideally this should have given exception.

---

Anyway, after seeing things like [this](https://github.com/ezsystems/ezplatform-form-builder/blob/master/src/lib/Tab/LocationView/SubmissionsTab.php#L206-L232) I decided it's best to skip this part of the change here and just continue to load normal core content class, for now. It will result in the same amount of SPI lookup calls in this case, so no harm done.

Updated, rebased and ready for new QA round.

micszo commented:
Not adding next label due to **TMP** commit!
